### ISSUE TITLE
CodeDeploy fails at "DownloadBundle" step

### ISSUE THREAD

**Issue by davidreid** [aws/aws-codedeploy-agent#255]
After the recent AWS CodeDeploy Agent update (version 1.1.1), all of my CodeDeploy deployments have failed at the "DownloadBundle" step.

In the console, this shows an error similar to: 

`Destination '/opt/codedeploy-agent/deployment-root/abcde123-4567-89edc-baabc-de1234567/d-ABCD1234/deployment-archive/appspec.yml' already exists`

This is occurring even when an instance is newly launched, with a fresh install of the CodeDeploy agent. Therefore, when Scaling Out an ASG, the scale-outs fail, if using the new version of the agent. Then, this causes a "scaling loop" to occur.

To rectify this, we have rolled out a downgrade of the agent to use version 1.1.0. This appears to resolve the issue.

From what I can see, this appears to be affecting Linux-based instances (Amazon Linux 1 and 2 from my testing) however, others may be affected as well.

**Comment by siarzukpiatrouski** [comment]
Same issue

**Comment by Helen1987** [comment]
As a workaround for your user data you can explicitly install version 1.1.0 of the Agent that does not have the issue. Check our docs:
```
sudo ./install auto -v releases/codedeploy-agent-1.1.0-4.noarch.rpm
```

**Comment by siarzukpiatrouski** [comment]
Thanks a lot, but this workaround was already described by the topic starter. Our auto-scaling configuration is too complex and it's a very difficult to apply changes on the go, so we can't use suggested fix. Is it possible  to provide estimates for this issue to be solved completely.

**Comment by brblck** [comment]
@siarzukpiatrouski can you share your deployment configuration with us?

**Comment by yubangxi** [comment]
@davidreid  @siarzukpiatrouski  sorry you ran into same issue. We are trying to reproduce the issue but have no luck. Do you mind sharing more information about your setup and help us answer some questions for us to better investigate and fix the issue?

* Are you using deployment revision from Github or S3?
* From the StackTrace you are using ZIP bundle type, can you confirm?
* Is this issue consistent on all instances?
* Would you mind sharing high level structure of your unzipped bundle?

Thanks!

**Comment by davidreid** [comment]
Hey @yubangxi 

This is what I have seen for far (I work for an MSP within the AWS Partner Network):

- This appears to affect Linux-based machines (so far I've seen this on Amazon Linux 1/2)
- All the instances are within an ASG and have Lifecycle Hooks into CodeDeploy 
- All the bundles are ZIP files
- I've seen this issue across ~8 AWS Accounts (with a fear this will affect more during a scale-out or deployment) and affecting a large number applications (all running within their own ASGs)
- For most of these applications, appspec.yml is in the root of the zip and further scripts (at each stage) to trigger Ansible, move our customers application into its respective directory and other bootstrapping etc.
- The bundle sizes vary massively from a few MBs to a couple hundred MB

Initially, we were installing the CodeDeploy Agent (through UserData) as per the docs. So, wget the file from S3, chmod +x and then ./install auto

Now, we have reverted the affected instances to perform a yum install of the version specific rpm. Upon using the older version, CodeDeploy works as expected.

What's strange to me is why CodeDeploy is failing at the DownloadBundle step, with the reason suggesting that the files (the agent itself has put there) shouldn't be there. It seems like the bundle is being downloaded and then the check to see if the directory is non-empty comes in afterwards?

This is happening on both freshly launched instances and existing instances (with previous deployments).

I hope this helps further?

**Comment by yubangxi** [comment]
@davidreid thanks for the response. A couple more questions:

* Will you be able to provide more context from CodeDeploy agent log, e.g. more lines above the error line? 
* >I've seen this issue across ~8 AWS Accounts
For those accounts you are seeing impact, with the same deployment bundle, is the deployment always failed with this error? Or the deployment did actually succeed sometimes?

With V1.1.1 update, one update we made is to use linux unzip command to unzip file instead of Ruby zip library when it's available, looks like in this case something wrong while running the unzip command. It will very helpful if you can try ``unzip -qo`` (the command agent is running) on you bundle and see anything is wrong.  We are not able to reproduce the issue so suspecting it's related specific bundle format/structure.

**Comment by siarzukpiatrouski** [comment]
Hi, @yubangxi 

We are using CloudFormation stacks for ASG, and OS is Amazon Linux 1. Bundle to deploy is a ZIP with TAR file in (for ex., /opt/codedeploy-agent/deployment-root/5e87205d-cf2f-4e66-a2bb-43c28742e8ba/d-8LB39XKI4/deployment-archive/html/magento.tar) 

We solved this issue by changing in stack template from:

`"command": "wget --quiet https://s3.amazonaws.com/aws-codedeploy-us-east-1/latest/install -O /tmp/codedeploy_install && /usr/bin/ruby /tmp/codedeploy_install auto",`

to:

`"command": "wget --quiet https://s3.amazonaws.com/aws-codedeploy-us-east-1/latest/install -O /tmp/codedeploy_install && /usr/bin/ruby /tmp/codedeploy_install auto -v releases/codedeploy-agent-1.1.0-4.noarch.rpm",`

<img width="1296" alt="Screen Shot 2020-07-11 at 00 58 17" src="https://user-images.githubusercontent.com/64413083/87206152-caa6a280-c311-11ea-9c23-d448675b7692.png">
<img width="1285" alt="Screen Shot 2020-07-11 at 00 58 32" src="https://user-images.githubusercontent.com/64413083/87206156-cc706600-c311-11ea-96a2-09ccf82a3013.png">

**Comment by davidreid** [comment]
@yubangxi 

> For those accounts you are seeing impact, with the same deployment bundle, is the deployment always failed with this error? Or the deployment did actually succeed sometimes?

Sorry, let me clarify. The deployment bundles are different between every account and every application. They failed consistently between all.

In regards to the CodeDeploy Agent logs, I have pulled these from earlier today, in CloudWatch: https://p.lee.io/9e683bae-8729-402a-9d6a-3237d660a4f5#encryptionKey=R0dr6sbqOeNe97UuZDPHKI3P

In regards to the `unzip -qo` - this looks to succeed with the same output as @siarzukpiatrouski

**Comment by Helen1987** [comment]
Agent version 1.1.1 is rolled back

**Comment by leytonreed** [comment]
> Agent version 1.1.1 is rolled back

Have attempted a scale out on an ASG using the method that was causing a problem:

```
wget https://aws-codedeploy-$REGION.s3.amazonaws.com/latest/install 
chmod +x ./install 
./install
```

...and can confirm it now installs version 1.1.0, and deployments occur without error.

**Comment by davidreid** [comment]
Agreed with @leytonreed 

The install script in S3 now installs version 1.1.0, and deployments succeed.

Similarly, I saw a number of "Action Required" sent to the various affected AWS Accounts, that had used version 1.1.1. However, the CodeDeploy Documentation doesn't make it clear that you have rolled back the version.

**Comment by AnandarajuCS** [comment]
@davidreid @siarzukpiatrouski Thanks for your information. As a followup to @yubangxi's questions, could you let us know how is the bundle zipped, in the sense, what commands have been used to zip the bundle?

**Comment by brblck** [comment]
@siarzukpiatrouski @davidreid @leytonreed thanks for your help in assisting our diagnosis of the issue. 

After much analysis, while it looks similar to a previous CodeDeploy agent issue, its actually a much more nuanced edge case that's resulting in a similar behavior. 

The core issue is that the agent is running into malformed / non-standard zip binaries during the `DownloadBundle` step and its resulting in a partial unpack or full failure of the unpack. When the scenario occurs, we have some failover logic that attempts to salvage the situation and that logic was actually generating a failure under these specific conditions.

We believe we have a good solution to the issue and we're working to prepare a v1.1.2 release of the agent that addresses it. 

**Would any of you be interested in helping us test a pre-release of v1.1.2 to validate that it addresses the issue you from your perspective?**

**Comment by davidreid** [comment]
@brblck - I certainly would be interested in testing the pre-release of v1.1.2, yes.

In regards to your comment, and @AnandarajuCS, the .zip files in this case were/are generated by Atlassian Bamboo so, it's interesting to hear that the deployment bundles could be malformed/non-standard. 

Aside from this, I'm certainly happy to test!

**Comment by brblck** [comment]
@davidreid thanks so much. The team is working to prepare a pre-release and validation instructions as we speak.

**Comment by AnandarajuCS** [comment]
This issue is fixed in our latest release v1.1.2 
https://aws.amazon.com/about-aws/whats-new/2020/08/aws-codedeploy-agent-improved-compatibility-amazon-linux-windows-ubuntu/
